tests/basic/test-database.cpp

```diff
@@ -0,0 +1,121 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include <geode/basic/bitsery_archive.h>
+#include <geode/basic/database.h>
+#include <geode/basic/identifier.h>
+#include <geode/basic/logger.h>
+#include <geode/basic/uuid.h>
+
+#include <geode/tests/common.h>
+
+struct Foo : public geode::Identifier
+{
+    Foo() = default;
+    Foo( double value ) : value_( value ) {}
+
+    template < typename Archive >
+    void serialize( Archive& archive )
+    {
+        archive.ext( *this, bitsery::ext::BaseClass< geode::Identifier >{} );
+        archive.value8b( value_ );
+    }
+
+    double value_{ 0 };
+};
+
+void register_foo_serializer( geode::PContext& context )
+{
+    context
+        .registerSingleBaseBranch< geode::Serializer, geode::Identifier, Foo >(
+            "Foo" );
+}
+
+void register_foo_deserializer( geode::PContext& context )
+{
+    context.registerSingleBaseBranch< geode::Deserializer, geode::Identifier,
+        Foo >( "Foo" );
+}
+
+void test_take_data( geode::Database& database, const geode::uuid& id )
+{
+    auto unique_id = database.take_data( id );
+    const auto& stolen_foo = dynamic_cast< Foo& >( *unique_id );
+    OPENGEODE_EXCEPTION(
+        stolen_foo.value_ == 42, "[Test] Wrong value after take data" );
+    auto foo_data = database.get_data( id );
+    const auto& foo = foo_data.get< Foo >();
+    OPENGEODE_EXCEPTION(
+        foo.value_ == 42, "[Test] Wrong value after register data" );
+    OPENGEODE_EXCEPTION( unique_id.get() != &foo,
+        "[Test] Objects adresses should be different" );
+}
+
+geode::uuid test_register_data( geode::Database& database )
+{
+    auto foo = database.register_data( Foo{ 42 } );
+    auto foo_data = database.get_data( foo );
+    OPENGEODE_EXCEPTION( foo_data.get< Foo >().value_ == 42,
+        "[Test] Wrong value after register data" );
+    return foo;
+}
+
+geode::uuid test_register_unique_data( geode::Database& database )
+{
+    auto foo = database.register_data( absl::make_unique< Foo >( 22 ) );
+    auto foo_data = database.get_data( foo );
+    OPENGEODE_EXCEPTION( foo_data.get< Foo >().value_ == 22,
+        "[Test] Wrong value after register unique data" );
+    return foo;
+}
+
+void test_modify_data( geode::Database& database, const geode::uuid& id )
+{
+    auto foo_data = database.get_data( id );
+    const auto& foo = foo_data.get< Foo >();
+    auto unique_id = database.take_data( id );
+    auto& taken_foo = dynamic_cast< Foo& >( *unique_id );
+    taken_foo.value_ = 12;
+    OPENGEODE_EXCEPTION(
+        taken_foo.value_ == 12, "[Test] Wrong value after modify taken data" );
+    OPENGEODE_EXCEPTION(
+        foo.value_ == 42, "[Test] Wrong value after register data" );
+    database.register_data( std::move( unique_id ) );
+    auto foo_data2 = database.get_data( id );
+    const auto& foo2 = foo_data2.get< Foo >();
+    OPENGEODE_EXCEPTION(
```

Comment:
and if you test foo here ? what happens?

Reply:
foo is supposed to have the old value, Let's try!